spec/outputs/sns_spec.rb

```diff
@@ -2,17 +2,110 @@
 require "logstash/devutils/rspec/spec_helper"
 require 'logstash/outputs/sns'
 require 'logstash/event'
+require "logstash/plugin_mixins/aws_config"
+
+require "aws-sdk" # TODO: Why is this not automatically brought in by the aws_config plugin?
 
 describe LogStash::Outputs::Sns do
-  describe '#format_message' do
-    it 'should allow to output to sns with empty tags' do
-      event = LogStash::Event.new({ "message" => "42 is the answer" })
-      expect(LogStash::Outputs::Sns.format_message(event)).to match(/Tags:\s\n/m)
+  let(:arn) { "arn:aws:sns:us-east-1:999999999:logstash-test-sns-topic" }
+  let(:sns_subject) { "The Plain in Spain" }
+  let(:sns_message) { "That's where the rain falls, plainly." }
+  let(:mock_client) { double("Aws::SNS::Client") }
+  let(:instance) {
+    allow(Aws::SNS::Client).to receive(:new).and_return(mock_client)
+    inst = LogStash::Outputs::Sns.new
+    allow(inst).to receive(:publish_boot_message_arn).and_return(nil)
+    inst.register
+    inst
+  }
+
+  describe "receiving an event" do
+    subject {
+      inst = instance
+      allow(inst).to receive(:send_sns_message).with(any_args)
+      inst.receive(event)
+      inst
+    }
+
+    shared_examples("publishing correctly") do
+      it "should send a message to the correct ARN if the event has 'arn' set" do
+        expect(subject).to have_received(:send_sns_message).with(arn, anything, anything)
+      end
+
+      it "should send the message" do
+        expect(subject).to have_received(:send_sns_message).with(anything, anything, expected_message)
+      end
+
+      it "should send the subject" do
+        expect(subject).to have_received(:send_sns_message).with(anything, sns_subject, anything)
+      end
     end
 
-    it 'should allow to output to sns with a list of tags' do
-      event = LogStash::Event.new({"tags" => ["elasticsearch", "logstash", "kibana"] })
-      expect(LogStash::Outputs::Sns.format_message(event)).to match(/\nTags:\selasticsearch,\slogstash,\skibana\n/m)
+    describe "with an explicit message" do
+      let(:expected_message) { sns_message }
+      let(:event) { LogStash::Event.new("sns" => arn, "sns_subject" => sns_subject, "sns_message" => sns_message) }
+      include_examples("publishing correctly")
     end
+
+    describe "without an explicit message" do
+      # Testing codecs sucks. It'd be nice if codecs had to implement some sort of encode_sync method
+      let(:expected_message) {
```

Comment:
Agree with this, I don't recall the history why we only have an async method. We don't have anything async.
This would make the code easier and cleaner IMHO.